File: Matlab files/pCpHdiagram.m
%Author:Riley Doyle
%Date: October 9 2020

%delete all variables in the workspace 
clear 

%delete all figures
close all

%delineate range of pHs from 2 to 14
pH_start=2;
pH_end= 14;

%define equilibrium constants 
T = 20 + 273.15;
S = 35;
Tc = 20;
P = 10; %(dbar)
t = Tc*1.00024;
p = P/10;
den = calc_density(S, t, p); %(kg/m3)
K_1 = calc_K1(T, S)*(den/1000);
K_2 = calc_K2(T, S)*(den/1000);
Kh = calc_Kh(T, S); %(mole/kg sol/atm)
Kh = Kh*(den/1000); %(mol/L/atm)
PCO2 = 0.000416; %(atm)

%create 100 linearly spaced points between pH_start and pH_end
vpH = linspace(pH_start, pH_end);

%length of vpH 
p = length(vpH);

%create an 100 by 1 array of zeros 
HCO3 = zeros(p,1);
CO3 = zeros(p,1);
CT = zeros(p,1);
H = zeros(p,1);
H2CO3 = zeros(p,1);

%for each point, calculate the following functions 
for i = 1:p
    H(i) = 10^(-vpH(i));
    H2CO3(i) = (Kh*PCO2);
    HCO3(i) = (((K_1*H2CO3(i))/H(i)));
    CO3(i) =(((K_2*K_2*H2CO3(i))/H(i)^2)); 
     CT(i) = H2CO3(i)+CO3(i)+HCO3(i);
end
H2CO3 = -log10(H2CO3);
HCO3 = -log10(HCO3);
CO3 = -log10(CO3);
CT = -log10(CT);

%interchange the rows and columns of vpH
vpH = transpose(vpH);
%concatenate arrays horizontally
carbonates = horzcat(vpH, H2CO3, HCO3, CO3, CT);

%modify r for plotting
%x axis is the first column of carbonates
x_axis = carbonates(:,1);
carbonates(:,1) = [];

%create a plot
figure(1)
set(gca, 'YDir', 'reverse');
plot(x_axis, carbonates)

%label x and y axis of the figure
xlabel('pH')
ylabel('pC')
legend('H_2CO_3', 'HCO_3^-', 'CO_3^-^2', 'C_T')


%delineate range of pHs from 2 to 14
pH_start=2;
pH_end= 14;

K_1 = calc_K1(T, S)*(den/1000);
pK1 = -log10(K_1);
K_2 = calc_K2(T, S)*(den/1000);
pK2 = -log10(K_2);

%create 100 linearly spaced points between pH_start and pH_end
vpH = linspace(pH_start, pH_end);

%length of vpH 
p = length(vpH);

%create an 100 by 1 array of zeros 
HCO3 = zeros(p,1);
CO3 = zeros(p,1);
CT = zeros(p,1);
H = zeros(p,1);
OH = zeros(p,1);
alpha0 = zeros(p,1);
alpha1 = zeros(p,1);
alpha2 = zeros(p,1);
H2CO3 = zeros(p,1);
alk = 2.5;

%for each point, calculate the following functions 
for i = 1:p
    H(i) = 10^(-vpH(i));
    OH(i) = 10^(-(14-vpH(i)));
    alpha0(i) = calc_alpha0( vpH(i), pK1, pK2);
    alpha1(i) = calc_alpha1( vpH(i), pK1, pK2);
    alpha2(i) = calc_alpha2( vpH(i), pK1, pK2);
    CT(i) = (alk - OH(i) + H(i))*(1/(alpha1(i) + (2*alpha2(i))));
    H2CO3(i) = (alpha0(i)*CT(i));
    HCO3(i) = (alpha1(i)*CT(i));
    CO3(i) = (alpha2(i)*CT(i)); 
end
H2CO3 = -log10(H2CO3);
HCO3 = -log10(HCO3);
CO3 = -log10(CO3);
CT = -log10(CT);
%interchange the rows and columns of vpH
vpH = transpose(vpH);
%concatenate arrays horizontally
carbonates = horzcat(vpH, H2CO3, HCO3, CO3, CT);

%modify r for plotting
%x axis is the first column of carbonates
x_axis = carbonates(:,1);
carbonates(:,1) = [];

%create a plot
figure(2)
set(gca, 'YDir', 'reverse');
plot(x_axis, carbonates)

%label x and y axis of the figure
xlabel('pH')
ylabel('pC')
legend('H_2CO_3', 'HCO_3^-', 'CO_3^-^2', 'C_T')

File: Python files/pCpHdiagram.py
#author: Riley Doyle
#date: 10/6/20
#file: pCpHdiagram
#status: working

#clear all
from IPython import get_ipython
get_ipython().magic('reset -sf')

import numpy as np
from calc_density import *
from calc_Ks import *
from calc_alphas import *
from calc_CO2_loss import *
import matplotlib.pyplot as plt

T = 20 + 273.15 #kelvins
S = 35 #g/kg
Tc = 20; #celcius
P = 10; #(dbar)
t = Tc*1.00024;
p = P/10;
den = calc_density(S, t, p); #(kg/m3)
K1 = calc_K1(T,S)*(den/1000) #mol/L
pK1 = -np.log10(K1)
K2 = calc_K2(T,S)*(den/1000)#mol/L
pK2 = -np.log10(K2)
Kh = calc_Kh(T,S)*(den/1000) #mol/L/atm
PCO2 = 0.000416
alk = 2.5
                

pH_start = 0
pH_end = 14
vpH = np.linspace(pH_start, pH_end)
p = len(vpH)

H2CO3 = np.zeros([p,1])
HCO3 = np.zeros([p,1])
CO3 = np.zeros([p,1])

for n in range(0,p):
    alpha0 = calc_alpha0(vpH, pK1, pK2)
    alpha1 = calc_alpha1(vpH, pK1, pK2)
    alpha2 = calc_alpha2(vpH, pK1, pK2)
    CO2sat = PCO2*Kh*1000 #mole/m3
                
    H = 10**(-vpH)
    OH = 10**(-(14-vpH))
    bt = (1/(alpha1 + (2*alpha2)))
    tp = (alk - OH + H)
    CT = tp * bt
    H2CO3 = -np.log10(alpha0*CT)
    HCO3 = -np.log10(alpha1*CT)
    CO3 = -np.log10(alpha2*CT)
    CT = -np.log10(CT)

vpH = np.transpose(vpH)
carbonates = np.concatenate((H2CO3, HCO3, CO3))

x_axis = vpH

fig = plt.figure()
ax = plt.subplot(111)
ax.plot(x_axis, H2CO3, label='H$_{2}$CO$_{3}$')
ax.plot(x_axis, HCO3, 'r', label='HCO$_3$$^{-}$')
ax.plot(x_axis, CO3, 'k', label='CO$_3$$^{-2}$')
ax.plot(x_axis, CT, label='C$_{t}$')
ax.set_xticks([2,4,6,8,10,12,14])
ax.set_ylim([14,-6])
ax.legend()
plt.xlabel('pH')
plt.ylabel('pC')
plt.show()
